File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/VB/Direct2D/D2DShapes/TextShape.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.

Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Globalization
Imports Microsoft.WindowsAPICodePack.DirectX.Direct2D1
Imports Microsoft.WindowsAPICodePack.DirectX.DirectWrite
Imports System.Text

Namespace D2DShapes
	Friend Class TextShape
		Inherits DrawingShape
		Private layoutRect As RectF
        Protected Friend dwriteFactory As DWriteFactory
        Protected Friend NiceGabriola As Boolean

		Private privateRenderingParams As RenderingParams
		<TypeConverter(GetType(ExpandableObjectConverter))> _
		Public Property RenderingParams() As RenderingParams
			Get
				Return privateRenderingParams
			End Get
			Set(ByVal value As RenderingParams)
				privateRenderingParams = value
			End Set
		End Property

		Private privateOptions? As DrawTextOptions
		<TypeConverter(GetType(ExpandableObjectConverter))> _
		Public Property Options() As DrawTextOptions?
			Get
				Return privateOptions
			End Get
			Set(ByVal value? As DrawTextOptions)
				privateOptions = value
			End Set
		End Property

		Private privateTextFormat As TextFormat
		<TypeConverter(GetType(ExpandableObjectConverter))> _
		Public Property TextFormat() As TextFormat
			Get
				Return privateTextFormat
			End Get
			Set(ByVal value As TextFormat)
				privateTextFormat = value
			End Set
		End Property

		Private privateText As String
		Public Property Text() As String
			Get
				Return privateText
			End Get
			Set(ByVal value As String)
				privateText = value
			End Set
		End Property

		Public Sub New(ByVal initialRenderTarget As RenderTarget, ByVal random As Random, ByVal d2DFactory As D2DFactory, ByVal bitmap As D2DBitmap, ByVal dwriteFactory As DWriteFactory)
			MyBase.New(initialRenderTarget, random, d2DFactory, bitmap)
			Me.dwriteFactory = dwriteFactory
            layoutRect = RandomRect(CanvasWidth, CanvasHeight)
            NiceGabriola = random.NextDouble() < 0.25 And dwriteFactory.SystemFontFamilyCollection.Contains("Gabriola")
			TextFormat = dwriteFactory.CreateTextFormat(RandomFontFamily(), RandomFontSize(), RandomFontWeight(), RandomFontStyle(), RandomFontStretch(), System.Globalization.CultureInfo.CurrentUICulture)
			If CoinFlip Then
				TextFormat.LineSpacing = RandomLineSpacing(TextFormat.FontSize)
			End If
			Text = RandomString(Random.Next(1000, 1000))

			FillBrush = RandomBrush()
			RenderingParams = RandomRenderingParams()

			If CoinFlip Then
				Options = DrawTextOptions.None
				If CoinFlip Then
					Options = Options Or DrawTextOptions.Clip
				End If
				If CoinFlip Then
					Options = Options Or DrawTextOptions.NoSnap
				End If
			End If
		End Sub

		Protected Friend Function RandomRenderingParams() As RenderingParams
			Dim rp As RenderingParams = dwriteFactory.CreateCustomRenderingParams(CSng(Math.Max(0.001, Math.Min(1, Random.NextDouble() * 2))), CSng(Math.Max(0, Math.Min(1, Random.NextDouble() * 3 - 1))), CSng(Math.Max(0, Math.Min(1, Random.NextDouble() * 3 - 1))), RandomPixelGeometry(), RandomRenderingMode())
			Return rp
		End Function

		Private Function RandomPixelGeometry() As PixelGeometry
			Return CType(Random.Next(0, 2), PixelGeometry)
		End Function

		Private Function RandomRenderingMode() As RenderingMode
			Return CType(Random.Next(0, 6), RenderingMode)
		End Function

		Protected Friend Function RandomLineSpacing(ByVal fontSize As Single) As LineSpacing
			Dim method As LineSpacingMethod = If(CoinFlip, LineSpacingMethod.Default, LineSpacingMethod.Uniform)
			Dim spacing = CSng(Random.NextDouble()*fontSize*4 + 0.5)
			Dim baseline = CSng(Random.NextDouble())
			Return New LineSpacing(method, spacing, baseline)
		End Function

		Private Function RandomFontFamily() As String
            If (NiceGabriola) Then
                Return "Gabriola"
            End If
            If CoinFlip Then
                'get random font out of the list of installed fonts
                Dim i As Integer = Random.Next(0, dwriteFactory.SystemFontFamilyCollection.Count - 1)
                Dim f As FontFamily = dwriteFactory.SystemFontFamilyCollection(i)
                Dim ret As String = Nothing
                If (f.FamilyNames.ContainsKey(CultureInfo.CurrentUICulture)) Then
                    ret = f.FamilyNames(CultureInfo.CurrentUICulture)
                ElseIf (f.FamilyNames.ContainsKey(CultureInfo.InvariantCulture)) Then
                    ret = f.FamilyNames(CultureInfo.InvariantCulture)
                ElseIf (f.FamilyNames.ContainsKey(CultureInfo.GetCultureInfo("EN-us"))) Then
                    ret = f.FamilyNames(CultureInfo.GetCultureInfo("EN-us"))
                Else
                    For Each c In f.FamilyNames.Keys
                        ret = f.FamilyNames(c)
                        Exit For
                    Next
                End If

                f.Dispose()
                Return ret
            End If
            'get one of the common fonts
            Return New String() {"Arial", "Times New Roman", "Courier New", "Impact", "Tahoma", "Calibri", "Consolas", "Segoe", "Cambria"}(Random.Next(0, 8))
        End Function

		Private Function RandomFontSize() As Single
			Return 6 + CSng(138 * Random.NextDouble() * Random.NextDouble())
		End Function

		Private Function RandomFontWeight() As FontWeight
			Return CType(Math.Min(950, 100 * Random.Next(1, 10)), FontWeight)
		End Function

		Private Function RandomFontStyle() As FontStyle
			Return CType(Random.Next(0, 2), FontStyle)
		End Function

		Private Function RandomFontStretch() As FontStretch
			Return CType(Random.Next(1, 9), FontStretch)
		End Function

		Private Function RandomString(ByVal size As Integer) As String
			Dim builder = New StringBuilder(size + 1) With {.Length = size}
			builder(0) = Convert.ToChar(Convert.ToInt32(Math.Floor(26 * Random.NextDouble() + 65)))
			For i As Integer = 1 To size - 2
				builder(i) = If(Random.NextDouble() < 0.2, " "c, Convert.ToChar(Convert.ToInt32(Math.Floor(26 * Random.NextDouble() + 97))))
			Next i
			builder(size - 1) = "."c
			Return builder.ToString()
		End Function

		Protected Friend Overrides Sub ChangeRenderTarget(ByVal newRenderTarget As RenderTarget)
			FillBrush = CopyBrushToRenderTarget(FillBrush, newRenderTarget)
		End Sub

		Protected Friend Overrides Sub Draw(ByVal renderTarget As RenderTarget)
			Dim stateBlock As DrawingStateBlock = d2DFactory.CreateDrawingStateBlock()
			renderTarget.SaveDrawingState(stateBlock)
            renderTarget.TextRenderingParams = RenderingParams

			If Options.HasValue Then
				renderTarget.DrawText(Text, TextFormat, layoutRect, FillBrush, Options.Value)
			Else
				renderTarget.DrawText(Text, TextFormat, layoutRect, FillBrush)
			End If

			renderTarget.RestoreDrawingState(stateBlock)
			stateBlock.Dispose()
		End Sub

		Public Overrides Function HitTest(ByVal point As Point2F) As Boolean
			Return point.X >= layoutRect.Left AndAlso point.Y >= layoutRect.Top AndAlso point.X <= layoutRect.Right AndAlso point.Y <= layoutRect.Bottom
		End Function

		Public Overrides Sub Dispose()
			If TextFormat IsNot Nothing Then
				TextFormat.Dispose()
			End If
			TextFormat = Nothing
			If RenderingParams IsNot Nothing Then
				RenderingParams.Dispose()
			End If
			RenderingParams = Nothing
			MyBase.Dispose()
		End Sub
	End Class
End Namespace

File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/CS/Direct2D/D2DShapes/TextShape.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using System.ComponentModel;
using Microsoft.WindowsAPICodePack.DirectX.Direct2D1;
using Microsoft.WindowsAPICodePack.DirectX.DirectWrite;
using System.Text;
using System.Globalization;

namespace D2DShapes
{
    internal class TextShape : DrawingShape
    {
        private RectF layoutRect;
        protected internal DWriteFactory dwriteFactory;

        [TypeConverter(typeof(ExpandableObjectConverter))]
        public RenderingParams RenderingParams { get; set; }

        [TypeConverter(typeof (ExpandableObjectConverter))]
        public DrawTextOptions? Options { get; set; }

        [TypeConverter(typeof(ExpandableObjectConverter))]
        public TextFormat TextFormat { get; set; }

        public string Text { get; set; }

        protected bool NiceGabriola;

        public TextShape(RenderTarget initialRenderTarget, Random random, D2DFactory d2DFactory, D2DBitmap bitmap, DWriteFactory dwriteFactory)
            : base(initialRenderTarget, random, d2DFactory, bitmap)
        {
            this.dwriteFactory = dwriteFactory;
            layoutRect = RandomRect(CanvasWidth, CanvasHeight);
            NiceGabriola = Random.NextDouble() < 0.25 && dwriteFactory.SystemFontFamilyCollection.Contains("Gabriola");
            TextFormat = dwriteFactory.CreateTextFormat(
                RandomFontFamily(),
                RandomFontSize(),
                RandomFontWeight(),
                RandomFontStyle(),
                RandomFontStretch(),
                System.Globalization.CultureInfo.CurrentUICulture);
            if (CoinFlip)
                TextFormat.LineSpacing = RandomLineSpacing(TextFormat.FontSize);
            Text = RandomString(Random.Next(1000, 1000));

            FillBrush = RandomBrush();
            RenderingParams = RandomRenderingParams();

            if (CoinFlip)
            {
                Options = DrawTextOptions.None;
                if (CoinFlip)
                    Options |= DrawTextOptions.Clip;
                if (CoinFlip)
                    Options |= DrawTextOptions.NoSnap;
            }
        }

        protected internal RenderingParams RandomRenderingParams()
        {
            RenderingParams rp = dwriteFactory.CreateCustomRenderingParams(
                (float)Math.Max(0.001, Math.Min(1, Random.NextDouble() * 2)), //gamma needs to be nonzero
                (float)Math.Max(0, Math.Min(1, Random.NextDouble() * 3 - 1)), //equal chances for 0, 1 and something in between
                (float)Math.Max(0, Math.Min(1, Random.NextDouble() * 3 - 1)), //equal chances for 0, 1 and something in between
                RandomPixelGeometry(),
                RandomRenderingMode());
            return rp;
        }

        private PixelGeometry RandomPixelGeometry()
        {
            return (PixelGeometry) Random.Next(0, 2);
        }

        private RenderingMode RandomRenderingMode()
        {
            return (RenderingMode)Random.Next(0, 6);
        }

        protected internal LineSpacing RandomLineSpacing(float fontSize)
        {
            LineSpacingMethod method = CoinFlip
                                           ? LineSpacingMethod.Default
                                           : LineSpacingMethod.Uniform;
            var spacing = (float)(Random.NextDouble()*fontSize*4 + 0.5);
            var baseline = (float)Random.NextDouble();
            return new LineSpacing(
                method,
                spacing,
                baseline);
        }

        private string RandomFontFamily()
        {
            if (NiceGabriola)
                return "Gabriola";
            if (CoinFlip)
            {
                //get random font out of the list of installed fonts
                int i = Random.Next(0, dwriteFactory.SystemFontFamilyCollection.Count - 1);
                FontFamily f = dwriteFactory.SystemFontFamilyCollection[i];
                string ret = null;

                if (f.FamilyNames.ContainsKey(CultureInfo.CurrentUICulture))
                    ret = f.FamilyNames[CultureInfo.CurrentUICulture];
                else if (f.FamilyNames.ContainsKey(CultureInfo.InvariantCulture))
                    ret = f.FamilyNames[CultureInfo.InvariantCulture];
                else if (f.FamilyNames.ContainsKey(CultureInfo.GetCultureInfo("EN-us")))
                    ret = f.FamilyNames[CultureInfo.GetCultureInfo("EN-us")];
                else
                {
                    foreach (var c in f.FamilyNames.Keys)
                    {
                        ret = f.FamilyNames[c];
                        break;
                    }
                }
                f.Dispose();
                return ret;
            }
            //get one of the common fonts
            return new[] {"Arial",
                           "Times New Roman",
                           "Courier New",
                           "Impact",
                           "Tahoma",
                           "Calibri",
                           "Consolas",
                           "Segoe",
                           "Cambria"
            }[Random.Next(0, 8)];
        }

        private float RandomFontSize()
        {
            return 6 + (float)(138 * Random.NextDouble() * Random.NextDouble());
        }

        private FontWeight RandomFontWeight()
        {
            return (FontWeight)(Math.Min(950, 100 * Random.Next(1, 10)));
        }

        private FontStyle RandomFontStyle()
        {
            return (FontStyle)Random.Next(0, 2);
        }

        private FontStretch RandomFontStretch()
        {
            return (FontStretch) Random.Next(1, 9);
        }

        private string RandomString(int size)
        {
            var builder = new StringBuilder(size + 1) { Length = size };
            builder[0] = Convert.ToChar(Convert.ToInt32(Math.Floor(26 * Random.NextDouble() + 65)));
            for (int i = 1; i < size - 1; i++)
            {
                builder[i] = Random.NextDouble() < 0.2 ?
                    ' ' :
                    Convert.ToChar(Convert.ToInt32(Math.Floor(26 * Random.NextDouble() + 97)));
            }
            builder[size - 1] = '.';
            return builder.ToString();
        }

        protected internal override void ChangeRenderTarget(RenderTarget newRenderTarget)
        {
            FillBrush = CopyBrushToRenderTarget(FillBrush, newRenderTarget);
        }

        protected internal override void Draw(RenderTarget renderTarget)
        {
            DrawingStateBlock stateBlock = d2DFactory.CreateDrawingStateBlock();
            renderTarget.SaveDrawingState(stateBlock);
            renderTarget.TextRenderingParams = RenderingParams;

            if (Options.HasValue)
            {
                renderTarget.DrawText(Text, TextFormat, layoutRect, FillBrush, Options.Value);
            }
            else
                renderTarget.DrawText(Text, TextFormat, layoutRect, FillBrush);

            renderTarget.RestoreDrawingState(stateBlock);
            stateBlock.Dispose();
        }

        public override bool HitTest(Point2F point)
        {
            return point.X >= layoutRect.Left &&
                point.Y >= layoutRect.Top &&
                point.X <= layoutRect.Right &&
                point.Y <= layoutRect.Bottom;
        }

        public override void Dispose()
        {
            if (TextFormat != null)
                TextFormat.Dispose();
            TextFormat = null;
            if (RenderingParams != null)
                RenderingParams.Dispose();
            RenderingParams = null;
            base.Dispose();
        }
    }
}
